File: templates/trie.py
# Trie (前缀树)
# 用法：
#   trie = Trie()
#   trie.insert("apple")
#   trie.search("apple")      # True
#   trie.search("app")        # False
#   trie.starts_with("app")   # True


class Trie:
    def __init__(self):
        self.children: dict[str, 'Trie'] = {}
        self.is_end = False

    def insert(self, word: str) -> None:
        node = self
        for c in word:
            if c not in node.children:
                node.children[c] = Trie()
            node = node.children[c]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self._find(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self._find(prefix) is not None

    def _find(self, s: str) -> 'Trie | None':
        node = self
        for c in s:
            if c not in node.children:
                return None
            node = node.children[c]
        return node

File: templates/trie.go
// Trie (前缀树)
// 用法：
//   trie := NewTrie()
//   trie.Insert("apple")
//   trie.Search("apple")     // true
//   trie.Search("app")       // false
//   trie.StartsWith("app")   // true

package templates

type Trie struct {
	children map[byte]*Trie
	isEnd    bool
}

func NewTrie() *Trie {
	return &Trie{children: make(map[byte]*Trie)}
}

func (t *Trie) Insert(word string) {
	node := t
	for i := 0; i < len(word); i++ {
		c := word[i]
		if _, ok := node.children[c]; !ok {
			node.children[c] = NewTrie()
		}
		node = node.children[c]
	}
	node.isEnd = true
}

func (t *Trie) Search(word string) bool {
	node := t.find(word)
	return node != nil && node.isEnd
}

func (t *Trie) StartsWith(prefix string) bool {
	return t.find(prefix) != nil
}

func (t *Trie) find(s string) *Trie {
	node := t
	for i := 0; i < len(s); i++ {
		next, ok := node.children[s[i]]
		if !ok {
			return nil
		}
		node = next
	}
	return node
}
